AppCenter/AppCenter/Internals/Context/AuthToken/MSAuthTokenContext.m

```diff
@@ -53,16 +48,67 @@ + (void)resetSharedInstance {
   sharedInstance = nil;
 }
 
-- (void)setAuthToken:(NSString *)authToken withAccountId:(NSString *)accountId {
+- (void)setAuthToken:(nullable NSString *)authToken withAccountId:(nullable NSString *)accountId expiresOn:(nullable NSDate *)expiresOn {
   NSArray *synchronizedDelegates;
   BOOL isNewUser = NO;
   @synchronized(self) {
-    self.authToken = authToken;
-    isNewUser = ![self.homeAccountId isEqualToString:accountId];
-    self.homeAccountId = accountId;
+
+    // If a nil authToken is passed with non-nil paarmeters, reset them.
+    if (!authToken) {
+      accountId = nil;
+      expiresOn = nil;
+    }
+    NSMutableArray<MSAuthTokenInfo *> *authTokenHistory = [[self authTokenHistory] mutableCopy];
+
+    MSAuthTokenInfo *lastEntry = authTokenHistory.lastObject;
+    NSString *__nullable latestAuthToken = lastEntry.authToken;
+    NSString *__nullable latestAccountId = lastEntry.accountId;
+    NSDate *__nullable latestTokenEndTime = lastEntry.expiresOn;
+
+    // If new token doesn't differ from the last token of array - no need to add it to array.
+    if (latestAuthToken != nil && authToken != nil && [latestAuthToken isEqualToString:(NSString * __nonnull) authToken]) {
```

Comment:
I think you can simplify the condition like below.
```suggestion
    if (latestAuthToken != nil && [latestAuthToken isEqualToString:(NSString * __nonnull) authToken]) {
```

Reply:
BTW don't you also want to skip when both are `nil`?

Reply:
@annakocheshkova This one is important, please fix it before merge